src/lib/kdf/kdf1_iso18033/kdf1_iso18033.cpp

```diff
@@ -9,41 +10,33 @@
 
 #include <botan/exceptn.h>
 #include <botan/internal/fmt.h>
+#include <botan/internal/stl_util.h>
 
 namespace Botan {
 
-void KDF1_18033::kdf(uint8_t key[],
-                     size_t key_len,
-                     const uint8_t secret[],
-                     size_t secret_len,
-                     const uint8_t salt[],
-                     size_t salt_len,
-                     const uint8_t label[],
-                     size_t label_len) const {
-   if(key_len == 0) {
+void KDF1_18033::perform_kdf(std::span<uint8_t> key,
+                             std::span<const uint8_t> secret,
+                             std::span<const uint8_t> salt,
+                             std::span<const uint8_t> label) const {
+   if(key.empty()) {
       return;
    }
 
-   const size_t blocks_required = key_len / m_hash->output_length();
+   const size_t blocks_required = key.size() / m_hash->output_length();
+   BOTAN_ARG_CHECK(blocks_required < 0xFFFFFFFF, "KDF1-18033 maximum output length exceeeded");
```

Comment:
Shouldn't this be `blocks_required < 0xFFFFFFFE`?

Reply:
Judging from just the implementation, I would argue that the check currently in master is too tight. Each block is indexed using a 32-bit counter starting at 0. So, we can at most produce 2^32 blocks before the output stream repeats.

On a 64-bit system (for `size_t` to be big enough to take `block_length * 2^32` bytes) this should even be fine with `blocks_required <= 0xFFFFFFFF`. The hashing loop does not use the counter as abort condition, hence it is technically okay if it wraps after the last iteration.

(Acknowledging that this is a completely theoretical discussion and that I didn't consult the spec whether they actually define the upper bound to be two blocks lower explicitly.)